plugin-locationcomponent/src/main/java/com/mapbox/maps/plugin/locationcomponent/LocationPuckManager.kt

```diff
@@ -196,50 +213,57 @@ internal class LocationPuckManager(
-        val modelScaleConstant = 2.0.pow(MAX_ZOOM - MIN_ZOOM)
-        val modelScaleExpression = puck.modelScaleExpression
-        val scaleExpression = if (modelScaleExpression == null) {
+        locationLayerRenderer.styleScaling(get3DPuckScaleExpression(puck, lastMercatorScale))
+      }
+    }
+  }
+
+  private fun get3DPuckScaleExpression(puck: LocationPuck3D, mercatorScale: Double = 1.0): Value {
+    val modelScaleConstant = 2.0.pow(MAX_ZOOM - MIN_ZOOM)
+    val modelScaleExpression = puck.modelScaleExpression
+    return if (modelScaleExpression == null) {
+      Value(
+        arrayListOf(
+          Value("interpolate"),
+          Value(arrayListOf(Value("exponential"), Value(0.5))),
+          Value(arrayListOf(Value("zoom"))),
+          Value(MIN_ZOOM),
           Value(
             arrayListOf(
-              Value("interpolate"),
-              Value(arrayListOf(Value("exponential"), Value(0.5))),
-              Value(arrayListOf(Value("zoom"))),
-              Value(MIN_ZOOM),
+              Value("literal"),
               Value(
                 arrayListOf(
-                  Value("literal"),
-                  Value(
-                    arrayListOf(
-                      Value(modelScaleConstant * puck.modelScale[0].toDouble()),
-                      Value(modelScaleConstant * puck.modelScale[1].toDouble()),
-                      Value(modelScaleConstant * puck.modelScale[2].toDouble())
-                    )
-                  )
+                  Value(modelScaleConstant * puck.modelScale[0].toDouble() * mercatorScale),
+                  Value(modelScaleConstant * puck.modelScale[1].toDouble() * mercatorScale),
+                  Value(modelScaleConstant * puck.modelScale[2].toDouble() * mercatorScale)
                 )
-              ),
-              Value(MAX_ZOOM),
+              )
+            )
+          ),
+          Value(MAX_ZOOM),
+          Value(
+            arrayListOf(
+              Value("literal"),
               Value(
                 arrayListOf(
-                  Value("literal"),
-                  Value(
-                    arrayListOf(
-                      Value(puck.modelScale[0].toDouble()),
-                      Value(puck.modelScale[1].toDouble()),
-                      Value(puck.modelScale[2].toDouble())
-                    )
-                  )
+                  Value(puck.modelScale[0].toDouble() * mercatorScale),
+                  Value(puck.modelScale[1].toDouble() * mercatorScale),
+                  Value(puck.modelScale[2].toDouble() * mercatorScale)
                 )
               )
             )
           )
-        } else {
-          Value.fromJson(modelScaleExpression).take()
-        }
-        locationLayerRenderer.styleScaling(scaleExpression)
-      }
+        )
+      )
+    } else {
+      Value.fromJson(modelScaleExpression).take()
     }
   }
 
+  private fun mercatorScale(lat: Double): Double {
+    return 1.0 / cos(lat * Math.PI / 180.0)
```

Comment:
will we get `zero division exception` when lat = 2/PI?

Reply:
Actually we don't need to divide by 1.0, corrected the calculation.